### PR TITLE
test: Add Node.js 18.x to workflows

### PR DESCRIPTION
**Pull request by sigv**
refs: #6489

## Description

Node.js 14.x is retained as the default, just as it was before, considering it is Maintenance LTS set to EOL next year (2023-04-30).

This change adds Node.js 18.x in all the places where workflows previously were building with Node.js 16.x - either as addition to it where multiple version are provided, or as an upgrade where only one version is used.

This finishes up the bulk of #6489, however manual testing of compatibility should be conducted further before that.

### Security Considerations

N/A. This is a change to test scenarios, to support the Current LTS of Node.js.

### Documentation Considerations

Backwards compatibility should not be affected. As result of manual testing in #6489, it may be decided to inform users that Node.js 18.x is now officially supported. Upgrade of Node.js versions is breaking for compiled blobs, in which case users need to rebuild project when upgrading from Node.js 16.x to 18.x, similar to how they would do when upgrading from 14.x to 16.x.

### Testing Considerations

The test suite is expanded to support Node.js 18.x. Additional manual testing is out of scope for this PR and can be evaluated/discussed in #6489.

### CODE REVIEW

**Comment by sigv** [comment]
Be advised, Mergify should be adjusted to pick up on 18.x tests, and they should be also marked as "Expected"/"Required".

**Comment by sigv** [comment]
[`packages/solo` `yarn test`](https://github.com/Agoric/agoric-sdk/actions/runs/3566790896/jobs/5993766480) invocation seems to have timed out after a TypeError. 🤔

**Comment by sigv** [comment]
Running locally with Node.js v18.12.1, I seem to get test execution in `packages/solo`:

```plain
yarn run v1.22.19
$ ava

# connecting
Loading slog sender modules: @agoric/telemetry/src/flight-recorder.js
Loading slog sender modules: @agoric/telemetry/src/flight-recorder.js
2022-11-28T17:43:41.521Z launch-chain: Expected SwingSet kernel statistic cosmic_swingset_inbound_queue_length not found
2022-11-28T17:43:41.522Z launch-chain: Expected SwingSet kernel statistic cosmic_swingset_inbound_queue_add not found
2022-11-28T17:43:41.522Z launch-chain: Expected SwingSet kernel statistic cosmic_swingset_inbound_queue_remove not found
agoric: deploy: running /root/agoric-sdk/packages/wallet/api/deploy.js

SES_UNHANDLED_REJECTION: (TypeError#1)
TypeError#1: Cannot deliver getBridge to target; typeof target is undefined

CapTP test fixture exception: (RemoteError(error:captp:unknown#20001)#1)
RemoteError(error:captp:unknown#20001)#1: board does not have id: (a string)
  at WebSocket.<anonymous> (packages/solo/test/captp-fixture.js:60:11)
  at WebSocket.emit (node:events:513:28)
  at Receiver.emit (node:events:513:28)
  at Socket.emit (node:events:513:28)

CapTP test fixture exception: (RemoteRangeError(error:captp:unknown#20002)#2)
RemoteRangeError(error:captp:unknown#20002)#2: id is probably a typo, cannot verify CRC: (a string)
  at WebSocket.<anonymous> (packages/solo/test/captp-fixture.js:60:11)
  at WebSocket.emit (node:events:513:28)
  at Receiver.emit (node:events:513:28)
  at Socket.emit (node:events:513:28)

  ✔ home.board (7.7s)
  ✔ home.wallet - transfer funds to the feePurse (4.4s)
  ✔ home.wallet - receive zoe invite (17.4s)
    ℹ {
        contractRoot: '/root/agoric-sdk/packages/zoe/src/contracts/automaticRefund.js',
      }
  ✔ home.wallet - central issuer setup (2.6s)
  ✔ home.localTimerService makeNotifier (1.1s)
  ✔ home.localTimerService makeRepeater (895ms)
2022-11-28T17:47:58.091Z shutdown: Shutting down cleanly...
2022-11-28T17:47:58.092Z shutdown: Process terminated!
  ─

  6 tests passed
Done in 427.42s.
```

**Comment by sigv** [comment]
Will try to understand what could be different between the CI worker and my local installation, resulting in `packages/solo` hanging..

**Comment by sigv** [comment]
`EBADF: bad file descriptor, read` during [build (14.x)](https://github.com/Agoric/agoric-sdk/actions/runs/3575824369/jobs/6012857144) when rebasing on top of latest `master` patches. Still haven't found a way to repro this locally.

**Comment by sigv** [comment]
It looks like `test-home` invoking the CLI does not return back to the Ava testcases. Must not believe the chain is ready for accepting work.

**Comment by sigv** [comment]
Bug in `@endo/captp` doesn't feel likely.. but maybe something with `makeCapTP` in particular? Still doesn't explain why this would cause an issue on the CI and not locally.

**Comment by sigv** [comment]
Could someone check that the runner `ubuntu-latest` actually has all packages up-to-date?
My local tests are on a fully updated Ubuntu 22.04.1 LTS (Jammy).

**Comment by mhofman** [comment]
I can't look at this this week, maybe @michaelfig can lend a hand?

**Comment by sigv** [comment]
Running that test on a clean Ubuntu 20.04 (Focal) image deployed in Azure on D8s_v5 (8 vCPUs, 32G RAM) seems to work. This feels more and more like a runner-specific issue, but I am not sure why this one package's test would be affected. 😐 

Here is the minimal scenario:

```bash
curl -fsSL https://deb.nodesource.com/gpgkey/nodesource.gpg.key | gpg --dearmor | sudo tee /etc/apt/trusted.gpg.d/nodesource.gpg >/dev/null
echo "deb [signed-by=/etc/apt/trusted.gpg.d/nodesource.gpg] https://deb.nodesource.com/node_18.x $(lsb_release -s -c) main" | sudo tee /etc/apt/sources.list.d/nodesource.list >/dev/null

sudo apt update
sudo apt --yes dist-upgrade
sudo apt --yes install chromium-browser gcc g++ jq make nodejs perl

sudo corepack enable yarn

sudo wget -O /usr/local/go1.19.3.linux-amd64.tar.gz https://go.dev/dl/go1.19.3.linux-amd64.tar.gz
sudo tar -C /usr/local -xf /usr/local/go1.19.3.linux-amd64.tar.gz
eval $(echo 'export PATH=$PATH:/usr/local/go/bin' | sudo tee /etc/profile.d/golang.sh)

go version
# go version go1.19.3 linux/amd64
node --version
# v18.12.1
yarn --version
# 1.22.19

git clone -b maint/test-node-18.x https://github.com/sigv/agoric-sdk.git
cd agoric-sdk

yarn install
yarn build

cd packages/solo
yarn test

# 6 tests passed
# Done in 218.96s.
```

**Comment by sigv** [comment]
Slowly learning more about how this is interconnected. Right now it seems the Websocket connection in [packages/solo/test/captp-fixture.js](https://github.com/Agoric/agoric-sdk/blob/0b5df16f83629efb7cb48d54250139e082ed109c/packages/solo/test/captp-fixture.js#L92) keep failing and looping - resulting in the observable mix of dots (`.`) and some `o`s. On Node 14.x and 16.x it's a consistent line of `o`s - no waiting on connection.

[Latest run](https://github.com/Agoric/agoric-sdk/actions/runs/3585408992/jobs/6033453540) does imply that the service itself comes online - as signaled by "Deployed Wallet!" - just that the test scenario never kicks off.

---

I also noticed that on [Node 14.x](https://github.com/Agoric/agoric-sdk/actions/runs/3585408992/jobs/6033453160) the log output says
```plain
CapTP test fixture exception: (RemoteError(error:captp:http://localhost:7999#20001)#1)
RemoteError(error:captp:http://localhost:7999#20001)#1: board does not have id: (a string)
  at WebSocket.<anonymous> (packages/solo/test/captp-fixture.js:60:11)
  at WebSocket.emit (events.js:400:28)
  at Receiver.emit (events.js:400:28)
  at Socket.emit (events.js:400:28)
```

however, on [Node 16.x](https://github.com/Agoric/agoric-sdk/actions/runs/3585408992/jobs/6033453318) it does not list the origin and instead says "unknown" as
```plain
CapTP test fixture exception: (RemoteError(error:captp:unknown#20001)#1)
RemoteError(error:captp:unknown#20001)#1: board does not have id: (a string)
  at WebSocket.<anonymous> (packages/solo/test/captp-fixture.js:60:11)
  at WebSocket.emit (node:events:513:28)
  at Receiver.emit (node:events:513:28)
  at Socket.emit (node:events:513:28)
```

I am locally observing "unknown" on Node 18.x as well, but just seemed a bit strange.

**Comment by sigv** [comment]
Node.js 17.0.0 enables native DNS result priority, which can result in IPv6 being preferred in many scenarios. I wonder if the IPv6 stack is the issue for the CI runner!

Node.js 16.x tests report `Open CapTP connection to ws://127.0.0.1:7999/private/captp...`
while Node.js 18.x tests fail with `connect ECONNREFUSED ::1:7999`

I did not enable IPv6 stack on my test VMs, and when testing locally the Linux box was only hooked up to IPv4 too. The test runner must have a native IPv6 address and prefers to use it then!

**Comment by sigv** [comment]
Sorry for the messy investigation and various rebases.

@mhofman, @michaelfig, this is ready for review and potential merge.

File: packages/solo/test/captp-fixture.js
Comment (mhofman) [inline packages/solo/test/captp-fixture.js:47]: Interesting! I see node now prefers IPv6, but I'm still wondering why that wouldn't work. I guess it doesn't fallback to IPv4 if the server doesn't answer and for some reason we're not listening on IPv6?

File: .github/workflows/deployment-test.yml
Comment (mhofman) [inline .github/workflows/deployment-test.yml]: I'll force integration to test if 18 works for the deployment test, but we may want to not poke the bear and stick with 16 here for now.

Comment (mhofman) [reply .github/workflows/deployment-test.yml]: Yeah looks like the version of node 18 available on GH actions is incompatible with the ancient version of ubuntu we're still using for the deployment test. I think we have an issue somewhere about upgrading the ubuntu version, or at least GH action is getting insistent, so maybe we can take care of switching to node 18 then.

Comment (mhofman) [inline .github/workflows/deployment-test.yml]:
```suggestion
          node-version: 16.x
```

**Review by mhofman — CHANGES_REQUESTED** [review]
Thanks for investigative work you did on this PR. Once we revert to 16 on the deployment test I believe this can be merged.

Please also [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for your branch so mergify can rebase as necessary before merging.

File: packages/solo/test/captp-fixture.js
Comment (sigv) [reply packages/solo/test/captp-fixture.js:47]: Indeed - there is no way to fall back to `127.0.0.1` when it's trying to use `::1`. It fails the connection and loops back around trying `::1` again.

File: .github/workflows/deployment-test.yml
Comment (sigv) [reply .github/workflows/deployment-test.yml]: Ack.

File: packages/solo/test/captp-fixture.js
Comment (sigv) [reply packages/solo/test/captp-fixture.js:47]: As for not listening on IPv6 - Generally requesting listen and only providing port enables all IPv4 and IPv6. I did not investigate further, but I would assume that listen here happens explicitly on `127.0.0.1` and that in turn enables only IPv4.

Comment (mhofman) [reply packages/solo/test/captp-fixture.js:47]: Yeah afaik, there is no way to open a single listen socket for both IPv4 and IPv6 restricted to the respective loopback addresses. I believe you can bind to a specific interface without address, but that's very system dependent and I doubt node exposes that. Alternative is to open 2 listen sockets. I suppose another way may be to listen providing `localhost` and hope both the server and client use the same resolution strategies.

Really [Node should implement Happy Eyeballs](https://github.com/nodejs/node/issues/41625), exactly for these kind of cases.

**Review by mhofman — APPROVED** [review]

**Comment by sigv** [comment]
@mhofman, don't forget about marking the 18.x tests as "Expected"/"Required" so they are visible clearly in the GitHub UI, as well as adding them to Mergify checks!

**Comment by sigv** [comment]
And big thanks for the review! 🎉

**Comment by mhofman** [comment]
> don't forget about marking the 18.x tests as "Expected"/"Required"

Yup, planning on making them required, and removing the 14 ones from required

**Comment by mhofman** [comment]
> And big thanks for the review! 🎉

Thanks so much for your contribution.

**Pull request merged**
